### PR TITLE
[WIP] Implemented user search

### PR DESCRIPTION
**Pull request by dhruvsinghal**
@pengwynn I have implemented the user search. Please have a look.

### CODE REVIEW

File: octokit/search.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// A service to return search records
+type UserSearchService struct {
+	client *Client
+	URL    *url.URL
+}
+
+// Get a list of search results based on SearchService#URL
+func (g *UserSearchService) All() (userSearchResults UserSearchResults,
```
Comment (pengwynn) [inline octokit/search.go]: I think we can get by with one service with four methods. How would you feel about:

```go
type SearchService struct {
	client *Client
	URL    *url.URL
}

// Get a list of search results based on SearchService#URL
func (g *SearchService) UserSearch() (userSearchResults UserSearchResults,
	result *Result) {
	result = g.client.get(g.URL, &userSearchResults)
	return
}
```

We can then add `CodeSearch`, `RepositorySearch`, etc.

**Comment by dhruvsinghal** [comment]
PTAL.

Also, I'm just curious as to why the service is initialized to a single URL. Won't that prevent the service from being reused? Also, why is it the responsibility of the user code to fill in the placeholders in the URL? Shouldn't it be parameters to the respective service calls (like UserSearch, CodeSearch, etc.)?

**Comment by pengwynn** [comment]
> Also, I'm just curious as to why the service is initialized to a single URL. Won't that prevent the service from being reused?

@dhruvsinghal It's a tradeoff. By accepting URLs, the client promotes passing around Hypermedia link relations, like those we get from [the API root][].

> Shouldn't it be parameters to the respective service calls (like UserSearch, CodeSearch, etc.)?

I _think_ we've got the same params for all four search flavors: `q`, `page`, `sort`, and `order`. Those could be named parameters or a `SearchParameters` struct I suppose.

[the API root]: https://api.github.com/

File: octokit/search.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package octokit
+
+import (
+	"github.com/jingweno/go-sawyer/hypermedia"
+	"net/url"
+)
+
+var SearchURL = Hyperlink("search{/type}?q={query}{&page,per_page,sort,order}")
+
+func (c *Client) Searches(url *url.URL) (searches *SearchService) {
```
Comment (pengwynn) [inline octokit/search.go]: I think this should be `Search`, singular.

**Comment by pengwynn** [comment]
@dhruvsinghal This is looking good. I'm tempted to implement another search type on this same branch before merging, just to see how our pattern will emerge.

@jingweno: any feedback?

**Comment by dhruvsinghal** [comment]
@pengwynn Sure, I'll implement another search, or if you want I can implement all the four searches on this branch itself.

**Comment by dhruvsinghal** [comment]
@pengwynn PTAL. I've implemented issue search as well.

File: octokit/search_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package octokit
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUserSearchService_All(t *testing.T) {
```
Comment (pengwynn) [inline octokit/search_test.go]: I think this can become `TestSearchService_UserSearch`, or just `TestUserSearch`.

**Comment by dhruvsinghal** [comment]
@pengwynn @jingweno, I propose the following restructuring of the ```SearchService``` Go API:

* Given a ```Client``` object, obtain a ```SearchService``` object by calling ```Search()```
* Now, depending on the type of search (user, code, issues, or repositories) call the appropriate method e.g. ```UserSearch``` on the ```SearchService``` object.
* The specific search object will take in the query, order and sort parameters (or nil if either of the latter two is not specified). This function will be responsible for expanding the URL that is defined at the top of this package.

Does thais sound reasonable?

File: octokit/search.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+type CodeFile struct {
+	*hypermedia.HALResource
+
+	Name       string     `json:"name,omitempty"`
+	Path       string     `json:"path,omitempty"`
+	SHA        string     `json:"sha,omitempty"`
+	URL        Hyperlink  `json:"url,omitempty"`
+	Git_URL    Hyperlink  `json:"git_url,omitempty"`
```
Comment (jingweno) [inline octokit/search.go]: Minor, Gophers prefer [mixed caps](https://golang.org/doc/effective_go.html#mixed-caps) :), so `Git_URL` => `GitURL`, similar for `HTML_URL`.

File: octokit/search_test.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		testMethod(t, r, "GET")
+		respondWithJSON(w, loadFixture("user_search.json"))
+	})
+
+	url, err := SearchURL.Expand(map[string]interface{}{
+		"type":  "users",
+		"query": "dhruvsinghal"})
+	assert.NoError(t, err)
+
+	searchResults, result := client.Search(url).UserSearch()
```
Comment (jingweno) [inline octokit/search_test.go]: What do you think of naming the method to `client.Search(url).Users()`? It seems to read more fluently.

File: octokit/search.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+// Get the repository search results based on SearchService#URL
+func (g *SearchService) RepositorySearch() (
+	repositorySearchResults RepositorySearchResults, result *Result) {
+	result = g.client.get(g.URL, &repositorySearchResults)
+	return
+}
+
+// Get the code search results based on SearchService#URL
+func (g *SearchService) CodeSearch() (
+	codeSearchResults CodeSearchResults, result *Result) {
```
Comment (jingweno) [inline octokit/search.go]: Hmm...the intention doesn't look very good. Have you tried running `go fmt ./...`? Reference: https://golang.org/cmd/gofmt/

**Comment by jingweno** [comment]
In general it looks good except some minor issues. Cheers!

File: octokit/search.go (hunk 6 above)
Comment (dhruvsinghal) [inline octokit/search.go]: I did run gofmt, but apparent it didn't make any changes.

**Comment by dhruvsinghal** [comment]
@jingweno Thanks for your comments!

@jingweno and @pengwynn PTAL.

**Comment by jingweno** [comment]
:+1: :shipit:

**Comment by dhruvsinghal** [comment]
@jingweno Should I merge this branch into master?

**Comment by pengwynn** [comment]
Gonna go ahead and merge this since it follows the same pattern as the other API methods.

@jingweno I'd like to explore how we can better support query params in methods like this. /cc https://github.com/octokit/go-octokit/issues/72

**Comment by jingweno** [comment]
@pengwynn 

> I'd like to explore how we can better support query params in methods like this. /cc #72

Yeah, this is a miss in the current design. Let's chat more about it on #72.

**Pull request merged**